### DISCUSSION TITLE
Network's z0 storage

### DISCUSSION THREAD

**Discussion by FranzForstmayr** [scikit-rf/scikit-rf#703] [Ideas]
Currently `z0` is stored in a hidden `_z0` variable which is a scalar for simple Networks and a array of shape `[f,n]` for every other case.
If a calculation is done which involves the recalculation of `z0` it's almost every time an array.

There were some PR and issues recently regarding this topic and I would suggest to store the reference impedance in an array regardless of the 'complexity'. This could be done in `z0.setter`. 

I think we could reorganize the relevant code around `z0`, `z0.setter` and `fix_z0_shape`. The only disadvantage I see is the additional memory needed for large arrays, which is filled with fifties most of the time.

In addition to that it's easier to return a reference to the array, which can be modified like requested in #664.

**Comment by JAnderson419** [comment]
Perhaps related to this, while well explained in the [tutorial](https://scikit-rf.readthedocs.io/en/latest/tutorials/Networks.html), creating networks from z/y/etc parameters is a bit clunky. From s parameters, you can create the network directly. From z parameters you use a `from_z()` method. With other parameters you use `rf.y2s`, etc and then define the network from s parameters. Ideally, creating the network would be as easy as just passing the z,y, etc parameters to the Network constructor. I tried creating a network directly from y parameters today and it seems like part of the issue preventing this is z0:

```
freq = rf.Frequency.from_f(dfloc['freq (Hz)'].to_numpy(), unit='Hz')
y = np.zeros((len(freq), 1, 1), dtype=complex)
y[:, 0, 0] = dfloc['Admittance (S)'].to_numpy()
d = rf.Network(frequency=freq, unit='Hz', y=y, z0=50)
```

```
  File "<ipython-input-26-0e35201a1e8f>", line 4, in <module>
    d = rf.Network(frequency=freq, unit='Hz', y=y, z0=[50])
  File "C:\Users\ander906\PycharmProjects\scratch_work\venv\lib\site-packages\skrf\network.py", line 468, in __init__
    self.__setattr__(attr, kwargs[attr])
  File "C:\Users\ander906\PycharmProjects\scratch_work\venv\lib\site-packages\skrf\network.py", line 1026, in y
    self._s = y2s(value, self.z0, s_def=self.s_def)
  File "C:\Users\ander906\PycharmProjects\scratch_work\venv\lib\site-packages\skrf\network.py", line 5733, in y2s
    z0 = fix_z0_shape(z0, nfreqs, nports)
  File "C:\Users\ander906\PycharmProjects\scratch_work\venv\lib\site-packages\skrf\network.py", line 6376, in fix_z0_shape
    elif len(z0) == nports:
TypeError: len() of unsized object
```

This can be worked around by converting to s before the network call or just redefining the network after creation:

```
freq = rf.Frequency.from_f(dfloc['freq (Hz)'].to_numpy(), unit='Hz')
y = np.zeros((len(freq), 1, 1), dtype=complex)
y[:, 0, 0] = dfloc['Admittance (S)'].to_numpy()

d = rf.Network(frequency=freq, unit='Hz', s=y, z0=50)
d.y = d.s
```

**Reply by jhillairet** [reply]
> Perhaps related to this, while well explained in the [tutorial](https://scikit-rf.readthedocs.io/en/latest/tutorials/Networks.html), creating networks from z/y/etc parameters is a bit clunky. From s parameters, you can create the network directly. From z parameters you use a `from_z()` method. With other parameters you use `rf.y2s`, etc and then define the network from s parameters. Ideally, creating the network would be as easy as just passing the z,y, etc parameters to the Network constructor. I tried creating a network directly from y parameters today and it seems like part of the issue preventing this is z0:

I'm working on updating this part of the tutorial, since since the last release it is now possible to build a `Network` just by passing the z, y or abcd parameters as well (thanks to @Ttl ).

**Reply by JAnderson419** [reply]
I guess I missed that. It is hard to keep up with all the great new updates happening recently. 😄  I am just now benefitting from the work done in #637.

**Reply by jhillairet** [reply]
> I am just now benefitting from the work done in #637.

Great. Let me know your feedback concerning Q-factor. It is still time to make some change in the API before making a new release. I'll add some comments on the documentation for example, in particular on the best practices concerning the choice of the bandwidth for the Q-fitting.

**Reply by jhillairet** [reply]
> I guess I missed that. It is hard to keep up with all the great new updates happening recently. 😄 

The PR was https://github.com/scikit-rf/scikit-rf/pull/657

**Reply by jhillairet** [reply]
>since since the last release it is now possible to build a `Network` just by passing the z, y or abcd parameters as well 

Well, I was wrong, we haven't released a version including this PR yet ;)

**Comment by mhuser** [comment]
Let's take a simple network with a scalar z0 and apply some transform to it, e.g. `se2gmm`: the memory footprint will grow because of z0 turning to an array.
If the original z0 would always be an array, it would use more memory initially, but this quantity would remain constant whatever transform is performed, without any surprises from the initial situation.
On the other hand, current systems usually have a lot of memory and it would take many simple networks transformations to fill it.

**Comment by FranzForstmayr** [comment]
Exactly, the network will grow practically every time, as even the simplest operations will access `z0` once.
See this example:

```python
import skrf as rf
import pickle

netw = rf.Network("skrf/tests/ntwk1.s2p")

# Now just pickle the network and look at the size
print(len(pickle.dump(netw)))
# 7116

print(netw)  # z0 is accessed here
# 2-Port Network: 'ntwk1',  1.0-10.0 GHz, 91 pts, z0=[50.+0.j 50.+0.j]
print(len(pickle.dump(netw)))
# 10020
```

To be honest, in my opinon it's not worth the trouble to keep the opportunity to have a single `z0`.
In addition to that, most of the unit tests should be made with the scalar and the array `z0`, #704  is a good example.


I'll try to make a draft PR the next days.
